File: Assets/PixelPlanets/StellarBreaker/Scripts/Persistence/SaveBinder.cs
using System;
using StellarBreaker.Core;
using StellarBreaker.Gameplay;

namespace StellarBreaker.Persistence
{
    /// <summary>Maps GameSession ⇆ SaveState. Boss state is intentionally not saved
    /// (a boss stage restarts cleanly on load).</summary>
    public static class SaveBinder
    {
        public static SaveState Capture(GameSession s)
        {
            var st = new SaveState
            {
                stardust     = BigNumberData.From(s.Wallet.Stardust),
                relics       = BigNumberData.From(s.Prestige.Relics.Stardust),
                tapLevel     = s.TapUpgrade.Level,
                shipLevels   = CaptureShipLevels(s.Ships),
                artifactLevels = CaptureArtifactLevels(s.Artifacts),
                currentStage = s.Stage.CurrentStage,
                highestStage = s.Stage.HighestStage,
            };
            return st;
        }

        public static void Apply(GameSession s, SaveState st)
        {
            if (st == null) return;
            s.Stage.RestoreProgress(st.currentStage, st.highestStage);  // before Begin()
            s.Wallet.Set(st.stardust.To());
            s.Prestige.Relics.Set(st.relics.To());
            s.TapUpgrade.Reset(Math.Max(1, st.tapLevel));
            s.Ships.RestoreLevels(st.shipLevels);
            s.Artifacts.RestoreLevels(st.artifactLevels);
        }

        static int[] CaptureShipLevels(ShipService ships)
        {
            var arr = new int[ships.Count];
            for (int i = 0; i < ships.Count; i++) arr[i] = ships.LevelOf(i);
            return arr;
        }

        static int[] CaptureArtifactLevels(ArtifactService arts)
        {
            var arr = new int[arts.Count];
            for (int i = 0; i < arts.Count; i++) arr[i] = arts.LevelOf(i);
            return arr;
        }
    }
}

File: Telegram/TelegramApp/src/game/persistence/SaveBinder.ts
// Ported from Assets/PixelPlanets/StellarBreaker/Scripts/Persistence/SaveBinder.cs
import { BigNumber, toBigNumberData } from '../core/BigNumber'
import type { GameSession } from '../gameplay/GameSession'
import { DailyRewardService } from '../monetization/DailyRewardService'
import type { SaveState } from './SaveState'

/**
 * Maps GameSession <-> SaveState. Boss state is intentionally not saved
 * (a boss stage restarts cleanly on load).
 */
export function captureSave(s: GameSession): SaveState {
  return {
    version: 1,
    stardust: toBigNumberData(s.wallet.balance),
    relics: toBigNumberData(s.prestige.relics.balance),
    antimatter: toBigNumberData(BigNumber.Zero),
    tapLevel: s.tapUpgrade.level,
    shipLevels: captureShipLevels(s),
    artifactLevels: captureArtifactLevels(s),
    missionProgress: s.missions.captureProgress(),
    missionClaimed: s.missions.captureClaimed(),
    currentStage: s.stage.currentStage,
    highestStage: s.stage.highestStage,
    lastSaveUnixSeconds: Math.floor(Date.now() / 1000),
    lastDailyClaimUnixSeconds: Number.isFinite(s.daily.lastClaimDay) ? s.daily.lastClaimDay * DailyRewardService.SECONDS_PER_DAY : 0,
    dailyStreak: s.daily.streak,
    stats: { ...s.stats, deepestStage: Math.max(s.stats.deepestStage, s.stage.highestStage) },
    offlineCapBonusHours: s.boosts.offlineCapBonusHours,
    vipExpiresUnixSeconds: s.boosts.vipExpiresUnixSeconds,
  }
}

export function applySave(s: GameSession, st: SaveState): void {
  s.stage.restoreProgress(st.currentStage, st.highestStage) // before begin()
  s.wallet.set(new BigNumber(st.stardust.mantissa, st.stardust.exponent))
  s.prestige.relics.set(new BigNumber(st.relics.mantissa, st.relics.exponent))
  s.tapUpgrade.reset(Math.max(1, st.tapLevel))
  s.ships.restoreLevels(st.shipLevels)
  s.artifacts.restoreLevels(st.artifactLevels)
  s.missions.restoreProgress(st.missionProgress, st.missionClaimed)
  s.boosts.offlineCapBonusHours = st.offlineCapBonusHours ?? 0
  s.boosts.vipExpiresUnixSeconds = st.vipExpiresUnixSeconds ?? 0
  if (st.lastDailyClaimUnixSeconds > 0) {
    s.daily.restore(Math.floor(st.lastDailyClaimUnixSeconds / DailyRewardService.SECONDS_PER_DAY), st.dailyStreak)
  }
  if (st.stats) {
    Object.assign(s.stats, st.stats)
  } else {
    // Pre-profile save: seed what we can infer so profiles never show zeros for veterans.
    s.stats.deepestStage = st.highestStage
    if (st.lastSaveUnixSeconds > 0) s.stats.firstPlayedUnixSeconds = st.lastSaveUnixSeconds
  }
  s.stats.deepestStage = Math.max(s.stats.deepestStage, st.highestStage)
  // Pre-#pack-dedup save: deepestBossCleared wasn't tracked yet. Backfill the highest boss
  // stage at-or-below deepestStage (reaching any stage past a boss REQUIRES having cleared
  // it - see StageManager.notifyPlanetKilled) so a migrating veteran's already-earned packs
  // aren't reinterpreted as "0 unique bosses cleared" and reset to zero. The server's own
  // grant floor (pack_progress.bosses_granted) is the actual safety net against a double
  // grant either way - see db.mjs's grantPacksFromSave.
  if (!st.stats || st.stats.deepestBossCleared === undefined) {
    let backfilled = 0
    for (let stage = s.stats.deepestStage; stage >= 1; stage--) {
      if (s.stage.isBossStage(stage)) {
        backfilled = stage
        break
      }
    }
    s.stats.deepestBossCleared = backfilled
  }
}

function captureShipLevels(s: GameSession): number[] {
  return Array.from({ length: s.ships.count }, (_, i) => s.ships.levelOf(i))
}

function captureArtifactLevels(s: GameSession): number[] {
  return Array.from({ length: s.artifacts.count }, (_, i) => s.artifacts.levelOf(i))
}
